### DISCUSSION TITLE
Improve token generator's secureness

### DISCUSSION THREAD

**Discussion by falsandtru** [npm/feedback#146] [Website] [answered]
The tokens generated on npmjs.com can accidentally be scraped by browser extensions. Shadow dom improves generator's secureness by blocking scraping. This is fail-safe design. Let's protect the tokens on npmjs.com.

**Comment by ljharb** [comment]
I’m a bit confused how shadow dom would help - browser extensions have superpowers, they can log everything you type, etc - they’re basically identical to you. If you don’t trust an extension, the only option is to not use it.

**Comment by falsandtru** [comment]
Some browser extensions, such as language translators (maybe also the builtin translator of Chrome), send the current page contents. Basically the secrets on the page would be sent together. Shadow dom can prevent this accident. A lot of people needs language translators for web browsing. They has the risk of accidentally sending the secrets. Can you protect them from the accidents and the risk?

**Reply by ljharb** [reply]
What do they currently do with password fields? What about credit card info, billing addresses, etc?

it seems reasonable to mitigate accidental exposure if this is a viable technique (altho it’d need fallbacks for browsers that don’t support shadow dom, which might defeat the purpose), but for the user this seems like a bit of an unsolvable problem in the general sense.

**Reply by falsandtru** [reply]
I'm talking about tokens, not passwords. The newly generated token on https://www.npmjs.com/settings/USERNAME/tokens/ can be collected via `document.body.textContent`. I believe this is too vulnerable to accidents. The texts which can be collected via `.textContent` are casually sent to some server. Probably Chrome's builtin translator also sends such texts.

**Reply by falsandtru** [reply]
> What about credit card info, billing addresses, etc?

Couldn't checked because I'm not a paid user.

**Comment by MylesBorins** [comment, marked as answer]
Hey @falsandtru I've pinged our internal security team to get some insight. In general if you think you've discovered something that could be a security vulnerability we do suggest submitting the report privately via https://www.npmjs.com/support.

In this certain circumstance I don't think there is a need to privately report, but just wanted to mention it

**Reply by falsandtru** [reply]
Thanks.
